### PR TITLE
[SPARK-14075] Refactor MemoryStore to be testable independent of BlockManager

### PR DESCRIPTION
**Pull request by JoshRosen**
This patch refactors the `MemoryStore` so that it can be tested without needing to construct / mock an entire `BlockManager`.
- The block manager's serialization- and compression-related methods have been moved from `BlockManager` to `SerializerManager`.
- `BlockInfoManager`is now passed directly to classes that need it, rather than being passed via the `BlockManager`.
- The `MemoryStore` now calls `dropFromMemory` via a new `BlockEvictionHandler` interface rather than directly calling the `BlockManager`. This change helps to enforce a narrow interface between the `MemoryStore` and `BlockManager` functionality and makes this interface easier to mock in tests.
- Several of the block unrolling tests have been moved from `BlockManagerSuite` into a new `MemoryStoreSuite`.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/storage/BlockManagerManagedBuffer.scala [hunk 1]
```diff
@@ -29,19 +29,19 @@ import org.apache.spark.util.io.ChunkedByteBuffer
  * to the network layer's notion of retain / release counts.
  */
 private[storage] class BlockManagerManagedBuffer(
-    blockManager: BlockManager,
+  blockInfoManager: BlockInfoManager,
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerManagedBuffer.scala]: 2 spaces

File: core/src/test/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorterSuite.java [hunk 2]
```diff
@@ -60,6 +60,8 @@
   final TestMemoryManager memoryManager =
     new TestMemoryManager(new SparkConf().set("spark.memory.offHeap.enabled", "false"));
   final TaskMemoryManager taskMemoryManager = new TaskMemoryManager(memoryManager, 0);
+  final SerializerManager serializerManager =
+      new SerializerManager(new JavaSerializer(new SparkConf()), new SparkConf());
```
Comment (andrewor14) [inline core/src/test/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorterSuite.java]: unindent

**Comment by andrewor14** [comment]
LGTM, will merge once you make it compile :) and pass tests.

File: core/src/test/scala/org/apache/spark/storage/MemoryStoreSuite.scala [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+  test("safely unroll blocks through putIterator") {
+    val (memoryStore, blockInfoManager) = makeMemoryStore(12000)
+    val smallList = List.fill(40)(new Array[Byte](100))
+    val bigList = List.fill(40)(new Array[Byte](1000))
+    def smallIterator: Iterator[Any] = smallList.iterator.asInstanceOf[Iterator[Any]]
+    def bigIterator: Iterator[Any] = bigList.iterator.asInstanceOf[Iterator[Any]]
+    assert(memoryStore.currentUnrollMemoryForThisTask === 0)
+
+    def putIterator[T](
+      blockId: BlockId,
```
Comment (JoshRosen) [inline core/src/test/scala/org/apache/spark/storage/MemoryStoreSuite.scala]: indent

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by JoshRosen** [comment]
```
[error] Test org.apache.spark.util.collection.unsafe.sort.UnsafeExternalSorterSuite.forcedSpillingWithoutComparator failed: java.io.IOException: Stream is corrupted, took 0.072 sec
[error]     at org.apache.spark.io.LZ4BlockInputStream.refill(LZ4BlockInputStream.java:166)
[error]     at org.apache.spark.io.LZ4BlockInputStream.read(LZ4BlockInputStream.java:118)
[error]     at java.io.DataInputStream.readInt(DataInputStream.java:387)
[error]     at org.apache.spark.util.collection.unsafe.sort.UnsafeSorterSpillReader.<init>(UnsafeSorterSpillReader.java:57)
[error]     at org.apache.spark.util.collection.unsafe.sort.UnsafeSorterSpillWriter.getReader(UnsafeSorterSpillWriter.java:149)
[error]     at org.apache.spark.util.collection.unsafe.sort.UnsafeExternalSorter.getIterator(UnsafeExternalSorter.java:559)
[error]     at org.apache.spark.util.collection.unsafe.sort.UnsafeExternalSorterSuite.forcedSpillingWithoutComparator(UnsafeExternalSorterSuite.java:395)
[error]     ...
```

Alas, this looks like a legitimate failure.

**Comment by JoshRosen** [comment]
Jenkins retest this please

**Comment by JoshRosen** [comment]
This passed tests, so I'm going to merge it in order to unblock progress on my other patches.

**Pull request closed without merging**